lrlex/src/lib/ctbuilder.rs

```diff
@@ -78,11 +78,29 @@ pub enum RustEdition {
     Rust2021,
 }
 
+use proc_macro2::TokenStream;
+use quote::{ToTokens, TokenStreamExt};
+/* Quote by default quotes option values as an empty string for `None`,
+ * and the value itself for `Some`, where here we actually want the
+ * Some/None included.
+ *
+ * See: https://github.com/dtolnay/quote/issues/20 */
+struct QuoteOption<T>(Option<T>);
+
+impl<T: ToTokens> ToTokens for QuoteOption<T> {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        tokens.append_all(match self.0 {
+            Some(ref t) => quote! { ::std::option::Option::Some(#t) },
+            None => quote! { ::std::option::Option::None },
+        });
+    }
+}
+
 /// A `CTLexerBuilder` allows one to specify the criteria for building a statically generated
 /// lexer.
 pub struct CTLexerBuilder<'a, LexerTypesT: LexerTypes = DefaultLexerTypes<u32>>
 where
-    LexerTypesT::StorageT: Debug + Eq + Hash,
+    LexerTypesT::StorageT: Debug + Eq + Hash + ToTokens,
```

Comment:
To me this here, and the related bounds immediately below  this are the most controversial aspects to this patch,
and the part that we need to consider most carefully.

The other thing we could do here is change this `ToTokens` to `Display`,

Changing it to `ToTokens` kind of just enables one minor cleanup, the `QuoteOption(tok_id)` cleanup in [18f1d9d](https://github.com/softdevteam/grmtools/pull/494/commits/18f1d9d6238f998edc673c8c5361fbd140885b38)

I think it would definitely be fine to consider using `Display` here instead of `ToTokens` and dropping that last cleanup?

Reply:
One more thing to note here, which I had mentioned in a previous comment on usize, 
using `ToTokens` here also has the effect of including the type in the value e.g. this patch is currently encoding `StorageT` values as `Some (5u32))`, a change from `Some(5)` with `Debug` and `Display`.

Looking at the sources generated by the testsuite these `StorageT` values i'm seeing are *all* using `u32`.
So if we do go with `ToTokens` it would be good to try and exercise this in the testsuite in a way that uses another type here `u8`.

Reply:
I don't have a very strong opinion here, but I'm somewhat leaning towards liking the `1_u8` form of `ToTokens`.

I tested the codegen of `ToTokens` here, by changing the `lrpar/examples/calc_actions/build.rs` to use
`CTLexerBuilder::<DefaultLexerTypes<u8>>::new_with_lexemet()` and it makes sense to me that in theory it should be fine for `StorageT` to include the concrete type in literals, since kind of the point is that it has a concrete size known to the caller.

Other places where we rely on the *can be converted into a usize losslessly* we could potentially use a *newtype* around known integer types  to print literals without the type suffix.

Either way, since we didn't remove the `Debug` suffix, at worst we can go back to using `Debug`?

Reply:
I'm fine with having the type suffix for literals.

One thing I haven't fully understood is: what types is `ToTokens` implemented on? I assume it's implemented on all primitive integers by `quote`? In that sense, this isn't going to be much of a breaking change unless a user does something odd like pass a newtype here: that would force them to know about the `ToTokens` trait and `quote`. In that sense, `Display` would be preferable, but I take your point in https://github.com/softdevteam/grmtools/pull/494/commits/18f1d9d6238f998edc673c8c5361fbd140885b38 that `Display` means we end up having to do our own string escaping, which is a bit nasty. On that basis, I'm fine using `ToTokens`.

Reply:
So long as the integer types are there, I think we'll have covered nearly every sensible use that I can conceive of.

Reply:
The one thing I guess I would say is that for the specific case of `StorageT` we don't have to worry about string escaping at all, I think it is also probably the most stable of all format printing we could likely expect format printing to be.  So I don't think it is too far fetched to use `ToTokens` internally, but not add the bounds for `StorageT` and use `Debug` or `Display` just for that.

So to me `ToTokens` over `Debug` or `Display` for `StorageT` mostly revolves around whether we want to be able to include the type suffix in integer literals. It isn't a huge thing but maybe adds a little bit of type checking we don't currently get (for better or worse!).

Reply:
If our plan is to use `quote` (and friends) for all code generation inside grmtools, I think the `ToTokens` bound makes sense: we'll have bought into that ecosystem, and might as well take advantage of it, and make clear that we've done so to users. If we don't plan on going on all the way with `quote` then perhaps we should just fall back on `Display`.